### ISSUE TITLE
Only apply softmax to alpha on operations leads to state explosion.

### ISSUE THREAD

**Issue by leckie-chn** [quark0/darts#36]
Hi,

I'm currently trying to reproduce your experiment in tensorflow. What I've found during the implementation & debugging is that, your current implementation of continuous relaxation (formula 2 in your paper) leads to state explosion. You just apply softmax on different operators, not different previous states, which will make the last state value much larger than the initial one.

For example, suppose `s0 = 1`, and every mixed operator is somehow a identity operator. Then `s1` will be 1, `s2=s0 + s1=2`, `s3=s0+s1+s2=4` ... finally the value of the states will grow exponentially. What's worse, the very large hidden state will feed to the next timestep/batch as initial state, which leads to infinite large state value. 

But anyway, your code worked. I think the key is to shrink the state value each time it go through the gates or cells. I thought the secret sauce is shared batchnorm, but it failed. 

So could you please explained for me?  
Thanks!

**Comment by quark0** [comment]
I don't see why the states would explode—they are supposed to be of zero mean and unit variance with batchnorms. You may try other types of normalization strategies as well, which should also work.

**Comment by leckie-chn** [comment]
@quark0 Yes theoretically normalization strategies will work, but in practice the mean & variance values in norm layers have a great chance that they can't catch up with the speed that state values grow (cold start). Under such circumstances weights & hyper-parameters need to be carefully chosen. 
In general I think that this kind of design introduces extra instability to the training of this network, though can be dealt with carefully tuning & norm techs, still requires modification.

**Comment by leckie-chn** [comment]
Fixed the bug by tuning batch norm layer. But I still think that this is a problem if we want to apply DARTS net to other network architectures.
